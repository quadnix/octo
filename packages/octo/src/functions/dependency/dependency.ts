import { UnknownModel } from '../../app.type.js';
import { DiffAction } from '../diff/diff.js';

class DependencyBehavior {
  forAction: DiffAction;

  onAction: DiffAction;

  onField: string;

  toField: string;

  constructor(onField: string, onAction: DiffAction, toField: string, forAction: DiffAction) {
    this.onField = onField;
    this.onAction = onAction;
    this.toField = toField;
    this.forAction = forAction;
  }
}

export enum DependencyRelationship {
  CHILD = 'child',
  PARENT = 'parent',
}

export interface IDependency {
  behaviors: {
    forAction: DependencyBehavior['forAction'];
    onAction: DependencyBehavior['onAction'];
    onField: DependencyBehavior['onField'];
    toField: DependencyBehavior['toField'];
  }[];

  from: string;

  relationship?: {
    onField: DependencyBehavior['onField'];
    toField: DependencyBehavior['toField'];
    type: DependencyRelationship;
  };

  to: string;
}

export class Dependency {
  private readonly behaviors: DependencyBehavior[] = [];

  readonly from: UnknownModel;

  private relationship: { onField: string; toField: string; type: DependencyRelationship };

  readonly to: UnknownModel;

  constructor(from: UnknownModel, to: UnknownModel) {
    this.from = from;
    this.to = to;
  }

  private getBehaviorIndex(onField: string, onAction: DiffAction, toField: string, forAction: DiffAction): number {
    return this.behaviors.findIndex(
      (b) => b.onField === onField && b.onAction === onAction && b.toField === toField && b.forAction === forAction,
    );
  }

  addBehavior(onField: string, onAction: DiffAction, toField: string, forAction: DiffAction): void {
    if (!this.from.hasOwnProperty(onField) || !this.to.hasOwnProperty(toField)) {
      throw new Error('Invalid field name is not a property of given model!');
    }

    if (this.getBehaviorIndex(onField, onAction, toField, forAction) === -1) {
      this.behaviors.push(new DependencyBehavior(onField, onAction, toField, forAction));
    }
  }

  addChildRelationship(onField: string, toField: string): void {
    if (this.relationship) {
      throw new Error('Dependency relationship already exists!');
    }
    this.relationship = { onField, toField, type: DependencyRelationship.CHILD };
  }

  addParentRelationship(onField: string, toField: string): void {
    if (this.relationship) {
      throw new Error('Dependency relationship already exists!');
    }
    this.relationship = { onField, toField, type: DependencyRelationship.PARENT };
  }

  hasMatchingBehavior(onField?: string, onAction?: DiffAction, toField?: string, forAction?: DiffAction): boolean {
    return this.behaviors.some((b) => {
      const onFieldCondition = onField ? b.onField === onField : true;
      const onActionCondition = onAction ? b.onAction === onAction : true;
      const toFieldCondition = toField ? b.toField === toField : true;
      const forActionCondition = forAction ? b.forAction === forAction : true;
      return onFieldCondition && onActionCondition && toFieldCondition && forActionCondition;
    });
  }

  isChildRelationship(): boolean {
    return this.relationship && this.relationship.type === DependencyRelationship.CHILD;
  }

  isEqual(dependency: Dependency): boolean {
    return (
      this.from.getContext() === dependency.from.getContext() && this.to.getContext() === dependency.to.getContext()
    );
  }

  isParentRelationship(): boolean {
    return this.relationship && this.relationship.type === DependencyRelationship.PARENT;
  }

  getRelationship(): { onField: string; toField: string; type: DependencyRelationship } | undefined {
    return this.relationship;
  }

  removeBehavior(onField: string, onAction: DiffAction, toField: string, forAction: DiffAction): void {
    const index = this.getBehaviorIndex(onField, onAction, toField, forAction);
    if (index === -1) {
      throw new Error('Dependency behavior not found!');
    }

    this.behaviors.splice(index, 1);
  }

  synth(): IDependency {
    const behaviors: IDependency['behaviors'] = [];
    this.behaviors.forEach((b) => {
      behaviors.push({
        forAction: b.forAction,
        onAction: b.onAction,
        onField: b.onField,
        toField: b.toField,
      });
    });

    return {
      behaviors,
      from: this.from.getContext(),
      relationship: this.relationship
        ? {
            onField: this.relationship.onField,
            toField: this.relationship.toField,
            type: this.relationship.type,
          }
        : undefined,
      to: this.to.getContext(),
    };
  }

  toJSON(): object {
    return {
      from: this.from.getContext(),
      relationship: this.relationship
        ? {
            onField: this.relationship.onField,
            toField: this.relationship.toField,
            type: this.relationship.type,
          }
        : undefined,
      to: this.to.getContext(),
    };
  }

  static unSynth(from: UnknownModel, to: UnknownModel, dependency: IDependency): Dependency {
    const newDependency = new Dependency(from, to);

    dependency.behaviors.forEach((b) => {
      newDependency.addBehavior(b.onField, b.onAction, b.toField, b.forAction);
    });

    if (dependency.relationship) {
      newDependency.relationship = { ...dependency.relationship };
    }

    return newDependency;
  }
}
