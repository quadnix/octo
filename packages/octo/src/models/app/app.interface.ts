import type { App } from './app.model.js';

/**
 * {@link App} model interface.
 *
 * @group Model Interfaces
 */
export interface IApp {
  name: App['name'];
}
