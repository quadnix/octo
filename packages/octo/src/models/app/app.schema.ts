import { Validate } from '../../decorators/validate.decorator.js';
import { Schema } from '../../functions/schema/schema.js';

export class AppSchema {
  @Validate({ options: { maxLength: 64, minLength: 2, regex: /^[a-zA-Z][\w-]*[a-zA-Z0-9]$/ } })
  name = Schema<string>();
}
