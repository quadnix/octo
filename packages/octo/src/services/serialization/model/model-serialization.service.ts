import {
  type IUnknownModel,
  type ModelSerializedOutput,
  NodeType,
  type UnknownModel,
  type UnknownOverlay,
} from '../../../app.type.js';
import { Container } from '../../../functions/container/container.js';
import { EventSource } from '../../../decorators/event-source.decorator.js';
import { Factory } from '../../../decorators/factory.decorator.js';
import {
  AnchorRegistrationEvent,
  ModelDeserializedEvent,
  ModelRegistrationEvent,
  ModelSerializedEvent,
  OverlayRegistrationEvent,
} from '../../../events/index.js';
import { Dependency, type IDependency } from '../../../functions/dependency/dependency.js';
import type { ANode } from '../../../functions/node/node.abstract.js';
import type { AAnchor } from '../../../overlays/anchor.abstract.js';
import type { IAnchor } from '../../../overlays/anchor.interface.js';
import { OverlayDataRepository } from '../../../overlays/overlay-data.repository.js';
import type { IOverlay } from '../../../overlays/overlay.interface.js';
import { ObjectUtility } from '../../../utilities/object/object.utility.js';

export class ModelSerializationService {
  private MODEL_DESERIALIZATION_TIMEOUT_IN_MS = 5000;

  private readonly classMapping: { [key: string]: any } = {};

  private async _deserialize(
    serializedOutput: ModelSerializedOutput,
    freeze: boolean,
  ): Promise<{ overlays: UnknownOverlay[]; root: UnknownModel }> {
    const deReferencePromises: {
      [p: string]: [Promise<boolean>, (value: boolean) => void, (error: Error) => void, NodeJS.Timeout];
    } = {};
    const seen: { [p: string]: UnknownModel } = {};

    const deReferenceContext = async (context: string): Promise<UnknownModel> => {
      if (!seen[context]) {
        if (deReferencePromises[context]) {
          await deReferencePromises[context][0];
        } else {
          deReferencePromises[context] = [] as any;
          const promise = new Promise<boolean>((resolve, reject) => {
            deReferencePromises[context][1] = resolve;
            deReferencePromises[context][2] = reject;
          });
          deReferencePromises[context][0] = promise;

          deReferencePromises[context][3] = setTimeout(() => {
            deReferencePromises[context][2](new Error('DeReferencing model operation timed out!'));
          }, this.MODEL_DESERIALIZATION_TIMEOUT_IN_MS);
          await promise;
        }
      }

      return seen[context];
    };

    const deserializeModel = async (context: string): Promise<UnknownModel> => {
      const { className, model } = serializedOutput.models[context];
      const deserializationClass = this.classMapping[className];

      seen[context] = await deserializationClass.unSynth(model, deReferenceContext);
      seen[context]['context'] = context;

      if (deReferencePromises[context] !== undefined) {
        deReferencePromises[context][1](true);
        clearTimeout(deReferencePromises[context][3]);
      }

      return seen[context];
    };

    // Deserialize all models.
    const promiseToDeserializeModels: Promise<UnknownModel>[] = [];
    for (const context in serializedOutput.models) {
      promiseToDeserializeModels.push(deserializeModel(context));
    }
    await Promise.all(promiseToDeserializeModels);

    // Deserialize all anchors.
    for (const a of serializedOutput.anchors) {
      const { className, parent } = a;
      const deserializationClass = this.classMapping[className];

      const anchor = await deserializationClass.unSynth(deserializationClass, a, deReferenceContext);
      seen[parent.context].addAnchor(anchor);
    }

    // Deserialize all overlays.
    const overlays: UnknownOverlay[] = [];
    for (const { className, overlay } of serializedOutput.overlays) {
      const deserializationClass = this.classMapping[className];
      const newOverlay = await deserializationClass.unSynth(deserializationClass, overlay, deReferenceContext);
      overlays.push(newOverlay);
      seen[newOverlay.getContext()] = newOverlay;
    }

    // Deserialize all dependencies.
    for (const d of serializedOutput.dependencies) {
      const dependency = Dependency.unSynth(seen[d.from], seen[d.to], d);
      if (!seen[d.from].getDependencies().some((d) => d.from === dependency.from && d.to === dependency.to)) {
        seen[d.from]['dependencies'].push(dependency);
      }
    }

    if (freeze) {
      for (const model of Object.values(seen)) {
        ObjectUtility.deepFreeze(model);
      }
      for (const overlay of overlays) {
        ObjectUtility.deepFreeze(overlay);
      }
    }

    // If no dependencies to serialize, return the first seen model.
    const root =
      serializedOutput.dependencies.length > 0
        ? seen[serializedOutput.dependencies[0].from]
        : seen[Object.keys(seen)[0]];

    return { overlays, root };
  }

  @EventSource(ModelDeserializedEvent)
  async deserialize(
    serializedOutput: ModelSerializedOutput,
    { freeze = true }: { freeze?: boolean } = {},
  ): Promise<UnknownModel> {
    const { overlays: oldOverlays, root } = await this._deserialize(serializedOutput, freeze);

    // Refresh the overlay data repository.
    await Container.getInstance().get(OverlayDataRepository, { args: [true, oldOverlays] });

    return root;
  }

  @EventSource(AnchorRegistrationEvent)
  @EventSource(ModelRegistrationEvent)
  @EventSource(OverlayRegistrationEvent)
  registerClass(className: string, deserializationClass: any): void {
    if (this.classMapping[className]) {
      throw new Error(`Class "${className}" is already registered!`);
    }
    this.classMapping[className] = deserializationClass;
  }

  @EventSource(ModelSerializedEvent)
  async serialize(root: UnknownModel): Promise<ModelSerializedOutput> {
    const boundary = root.getBoundaryMembers();
    const anchors: (IAnchor & { className: string })[] = [];
    const dependencies: IDependency[] = [];
    const models: { [key: string]: { className: string; model: IUnknownModel } } = {};
    const overlays: { className: string; overlay: IOverlay }[] = [];

    for (const model of boundary) {
      for (const d of model.getDependencies()) {
        // Skip dependencies that are not part of boundary.
        if (
          !boundary.some((m) => m.getContext() === d.from.getContext()) ||
          !boundary.some((m) => m.getContext() === d.to.getContext())
        ) {
          continue;
        }

        dependencies.push(d.synth());
      }

      if ((model.constructor as typeof ANode).NODE_TYPE === NodeType.MODEL) {
        const context = model.getContext();
        if (!models[context]) {
          models[context] = {
            className: `${(model.constructor as typeof ANode).NODE_PACKAGE}/${model.constructor.name}`,
            model: model.synth() as IUnknownModel,
          };
        }

        for (const a of (model as UnknownModel).getAnchors()) {
          anchors.push({
            ...a.synth(),
            className: `${(a.constructor as typeof AAnchor).NODE_PACKAGE}/${a.constructor.name}`,
          });
        }
      } else if ((model.constructor as typeof ANode).NODE_TYPE === NodeType.OVERLAY) {
        overlays.push({
          className: `${(model.constructor as typeof ANode).NODE_PACKAGE}/${model.constructor.name}`,
          overlay: (model as UnknownOverlay).synth(),
        });
      }
    }

    return { anchors, dependencies, models, overlays };
  }

  setModelDeserializationTimeout(timeoutInMs: number): void {
    this.MODEL_DESERIALIZATION_TIMEOUT_IN_MS = timeoutInMs;
  }
}

@Factory<ModelSerializationService>(ModelSerializationService)
export class ModelSerializationServiceFactory {
  private static instance: ModelSerializationService;

  static async create(forceNew = false): Promise<ModelSerializationService> {
    if (!this.instance) {
      this.instance = new ModelSerializationService();
    }
    if (forceNew) {
      const newInstance = new ModelSerializationService();
      Object.keys(this.instance).forEach((key) => (this.instance[key] = newInstance[key]));
    }
    return this.instance;
  }
}
