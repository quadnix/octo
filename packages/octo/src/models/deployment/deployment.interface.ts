import { IModelReference } from '../model.interface.js';
import { Deployment } from './deployment.model.js';

export interface IDeployment {
  deploymentTag: Deployment['deploymentTag'];
  image: IModelReference;
}
