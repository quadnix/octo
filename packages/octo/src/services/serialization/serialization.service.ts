import { App } from '../../models/app/app.model';
import { Deployment } from '../../models/deployment/deployment.model';
import { Environment } from '../../models/environment/environment.model';
import { Execution } from '../../models/execution/execution.model';
import { Image } from '../../models/image/image.model';
import { Model } from '../../models/model.abstract';
import { IModel } from '../../models/model.interface';
import { Pipeline } from '../../models/pipeline/pipeline.model';
import { Region } from '../../models/region/region.model';
import { Server } from '../../models/server/server.model';
import { Support } from '../../models/support/support.model';
import { Dependency, IDependency } from '../../functions/dependency/dependency.model';

export type SerializedOutput = {
  dependencies: IDependency[];
  models: { [p: string]: { className: string; model: IModel<unknown, unknown> } };
  version: string;
};

export class SerializationService {
  readonly SERIALIZATION_VERSION = 'v0';

  private readonly classMapping: { [key: string]: any } = {
    App,
    Deployment,
    Environment,
    Execution,
    Image,
    Pipeline,
    Region,
    Server,
    Support,
  };

  private throwErrorIfDeserializationClassInvalid(deserializationClass: any): void {
    const isValid = typeof deserializationClass?.unSynth === 'function';
    if (!isValid) {
      throw new Error('Invalid class, no reference to unSynth static method!');
    }
  }

  async deserialize(serializedOutput: SerializedOutput): Promise<Model<unknown, unknown>> {
    if (serializedOutput.version !== this.SERIALIZATION_VERSION) {
      throw new Error('Version mismatch on deserialization!');
    }

    const deReferencePromises: { [p: string]: (value: boolean) => void } = {};
    const seen: { [p: string]: Model<unknown, unknown> } = {};

    const deReferenceContext = async (context: string): Promise<Model<unknown, unknown>> => {
      if (!seen[context]) {
        const promise = new Promise<boolean>((resolve) => {
          deReferencePromises[context] = resolve;
        });
        await promise;
      }

      return seen[context];
    };

    for (const d of serializedOutput.dependencies) {
      if (!seen[d.from]) {
        const { className, model } = serializedOutput.models[d.from];
        const deserializationClass = this.classMapping[className];
        this.throwErrorIfDeserializationClassInvalid(deserializationClass);

        seen[d.from] = await deserializationClass.unSynth(model, deReferenceContext);
        if (deReferencePromises[d.from] !== undefined) {
          deReferencePromises[d.from](true);
        }
      }

      if (!seen[d.to]) {
        const { className, model } = serializedOutput.models[d.to];
        const deserializationClass = this.classMapping[className];
        this.throwErrorIfDeserializationClassInvalid(deserializationClass);

        seen[d.to] = await deserializationClass.unSynth(model, deReferenceContext);
        if (deReferencePromises[d.to] !== undefined) {
          deReferencePromises[d.to](true);
        }
      }

      const dependency = Dependency.unSynth(seen[d.from], seen[d.to], d);
      seen[d.from]['dependencies'].push(dependency);
    }

    return seen[serializedOutput.dependencies[0].from];
  }

  registerClass(className: string, deserializationClass: any): void {
    this.throwErrorIfDeserializationClassInvalid(deserializationClass);
    this.classMapping[className] = deserializationClass;
  }

  serialize(root: Model<unknown, unknown>): SerializedOutput {
    const boundary = root.getBoundaryMembers();
    const dependencies: IDependency[] = [];
    const models: { [key: string]: { className: string; model: IModel<unknown, unknown> } } = {};

    for (const model of boundary) {
      for (const d of model['dependencies']) {
        // Skip dependencies that are not part of boundary.
        if (
          !boundary.some((m) => m.getContext() === d.from.getContext()) ||
          !boundary.some((m) => m.getContext() === d.to.getContext())
        ) {
          continue;
        }

        dependencies.push(d.synth());
      }

      const context = model.getContext();
      if (!models[context]) {
        models[context] = {
          className: model.constructor.name,
          model: model.synth() as IModel<unknown, unknown>,
        };
      }
    }

    return { dependencies, models, version: this.SERIALIZATION_VERSION };
  }
}
