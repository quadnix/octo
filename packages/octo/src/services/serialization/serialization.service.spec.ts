import { IDependency } from '../../functions/dependency/dependency.model';
import { App } from '../../models/app/app.model';
import { Deployment } from '../../models/deployment/deployment.model';
import { Environment } from '../../models/environment/environment.model';
import { Image } from '../../models/image/image.model';
import { Region } from '../../models/region/region.model';
import { Server } from '../../models/server/server.model';
import { Support } from '../../models/support/support.model';
import { SerializationService, SerializedOutput } from './serialization.service';

describe('Serialization Service UT', () => {
  describe('deserialize()', () => {
    it('should throw error when de-serializing with a different version', async () => {
      const serializationService = new SerializationService();
      await expect(async () => {
        await serializationService.deserialize({ version: 'v1' } as SerializedOutput);
      }).rejects.toThrowErrorMatchingInlineSnapshot(`"Version mismatch on deserialization!"`);
    });

    it('should throw error when de-serializing an unknown class', async () => {
      const serializedOutput: SerializedOutput = {
        dependencies: [
          {
            from: 'App',
          } as IDependency,
        ],
        models: {
          App: { className: 'ClassNotExist', model: null },
        } as any,
        version: 'v0',
      };

      const serializationService = new SerializationService();
      await expect(async () => {
        await serializationService.deserialize(serializedOutput);
      }).rejects.toThrowErrorMatchingInlineSnapshot(`"Invalid class, no reference to unSynth static method!"`);
    });

    it('should deserialize known classes', async () => {
      const app0 = new App('test-app');
      const region0 = new Region('region-0');
      app0.addRegion(region0);

      const serializationService = new SerializationService();
      serializationService.registerClass('App', App);

      const output = serializationService.serialize(app0);
      const app1 = (await serializationService.deserialize(output)) as App;

      expect(app1.name).toBe('test-app');
    });
  });

  describe('serialize()', () => {
    it('should serialize an empty app', () => {
      const app0 = new App('test-app');

      const serializationService = new SerializationService();
      expect(serializationService.serialize(app0)).toMatchSnapshot();
    });

    it('should serialize a non-empty app', () => {
      const app0 = new App('test-app');
      const region0 = new Region('region-1');
      const environment0 = new Environment('qa');
      app0.addRegion(region0);
      app0.addServer(new Server('backend'));
      app0.addSupport(new Support('nginx', 'nginx'));
      environment0.environmentVariables.set('key', 'value');
      region0.addEnvironment(environment0);

      const serializationService = new SerializationService();
      expect(serializationService.serialize(app0)).toMatchSnapshot();
    });

    it('should serialize only boundary members', () => {
      const app0 = new App('test-app');
      const region0_1 = new Region('region-1');
      const region0_2 = new Region('region-2');
      const environment0_1 = new Environment('qa');
      const environment0_2 = new Environment('qa');
      app0.addRegion(region0_1);
      app0.addRegion(region0_2);
      region0_1.addEnvironment(environment0_1);
      region0_2.addEnvironment(environment0_2);

      const serializationService = new SerializationService();
      expect(serializationService.serialize(region0_1)).toMatchSnapshot();
    });

    it('should throw error when trying to add same child to two parent', () => {
      const app0 = new App('test-app');
      const region0_0 = new Region('region-0');
      const region0_1 = new Region('region-1');
      const environment0 = new Environment('qa');
      region0_0.addEnvironment(environment0);
      region0_1.addEnvironment(environment0);
      app0.addRegion(region0_0);
      app0.addRegion(region0_1);

      const serializationService = new SerializationService();
      expect(() => serializationService.serialize(app0)).toThrowErrorMatchingInlineSnapshot(
        `"Found circular dependencies!"`,
      );
    });

    it('should serialize when multiple models have dependency on same model', async () => {
      const serializationService = new SerializationService();

      const app0 = new App('test-app');
      const image0 = new Image('backend-runner', '0.0.1', { dockerFilePath: '/Dockerfile' });
      const server0 = new Server('backend');
      const deployment0 = new Deployment('0.0.1', image0);
      const deployment1 = new Deployment('0.0.2', image0);
      app0.addImage(image0);
      app0.addServer(server0);
      server0.addDeployment(deployment0);
      server0.addDeployment(deployment1);

      expect(serializationService.serialize(app0)).toMatchSnapshot();
    });
  });
});
