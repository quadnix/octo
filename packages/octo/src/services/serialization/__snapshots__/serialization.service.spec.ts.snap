// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Serialization Service UT serialize() should serialize a non-empty app 1`] = `
{
  "dependencies": [
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "name",
          "toField": "regionId",
        },
      ],
      "from": "app=test-app",
      "relationship": {
        "onField": "name",
        "toField": "regionId",
        "type": "parent",
      },
      "to": "region=region-1,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "name",
          "toField": "serverKey",
        },
      ],
      "from": "app=test-app",
      "relationship": {
        "onField": "name",
        "toField": "serverKey",
        "type": "parent",
      },
      "to": "server=backend,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "name",
          "toField": "serverKey",
        },
      ],
      "from": "app=test-app",
      "relationship": {
        "onField": "name",
        "toField": "serverKey",
        "type": "parent",
      },
      "to": "support=nginx,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "serverKey",
          "toField": "name",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "serverKey",
          "toField": "name",
        },
      ],
      "from": "support=nginx,app=test-app",
      "relationship": {
        "onField": "serverKey",
        "toField": "name",
        "type": "child",
      },
      "to": "app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "serverKey",
          "toField": "name",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "serverKey",
          "toField": "name",
        },
      ],
      "from": "server=backend,app=test-app",
      "relationship": {
        "onField": "serverKey",
        "toField": "name",
        "type": "child",
      },
      "to": "app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "regionId",
          "toField": "name",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "regionId",
          "toField": "name",
        },
      ],
      "from": "region=region-1,app=test-app",
      "relationship": {
        "onField": "regionId",
        "toField": "name",
        "type": "child",
      },
      "to": "app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "regionId",
          "toField": "environmentName",
        },
      ],
      "from": "region=region-1,app=test-app",
      "relationship": {
        "onField": "regionId",
        "toField": "environmentName",
        "type": "parent",
      },
      "to": "environment=qa,region=region-1,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "environmentName",
          "toField": "regionId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "environmentName",
          "toField": "regionId",
        },
      ],
      "from": "environment=qa,region=region-1,app=test-app",
      "relationship": {
        "onField": "environmentName",
        "toField": "regionId",
        "type": "child",
      },
      "to": "region=region-1,app=test-app",
    },
  ],
  "models": {
    "app=test-app": {
      "className": "App",
      "model": {
        "name": "test-app",
      },
    },
    "environment=qa,region=region-1,app=test-app": {
      "className": "Environment",
      "model": {
        "environmentName": "qa",
        "environmentVariables": {
          "key": "value",
        },
      },
    },
    "region=region-1,app=test-app": {
      "className": "Region",
      "model": {
        "regionId": "region-1",
      },
    },
    "server=backend,app=test-app": {
      "className": "Server",
      "model": {
        "serverKey": "backend",
      },
    },
    "support=nginx,app=test-app": {
      "className": "Support",
      "model": {
        "applicationType": "nginx",
        "serverKey": "nginx",
      },
    },
  },
  "version": "v0",
}
`;

exports[`Serialization Service UT serialize() should serialize an empty app 1`] = `
{
  "dependencies": [],
  "models": {
    "app=test-app": {
      "className": "App",
      "model": {
        "name": "test-app",
      },
    },
  },
  "version": "v0",
}
`;

exports[`Serialization Service UT serialize() should serialize only boundary members 1`] = `
{
  "dependencies": [
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "regionId",
          "toField": "name",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "regionId",
          "toField": "name",
        },
      ],
      "from": "region=region-1,app=test-app",
      "relationship": {
        "onField": "regionId",
        "toField": "name",
        "type": "child",
      },
      "to": "app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "regionId",
          "toField": "environmentName",
        },
      ],
      "from": "region=region-1,app=test-app",
      "relationship": {
        "onField": "regionId",
        "toField": "environmentName",
        "type": "parent",
      },
      "to": "environment=qa,region=region-1,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "name",
          "toField": "regionId",
        },
      ],
      "from": "app=test-app",
      "relationship": {
        "onField": "name",
        "toField": "regionId",
        "type": "parent",
      },
      "to": "region=region-1,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "environmentName",
          "toField": "regionId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "environmentName",
          "toField": "regionId",
        },
      ],
      "from": "environment=qa,region=region-1,app=test-app",
      "relationship": {
        "onField": "environmentName",
        "toField": "regionId",
        "type": "child",
      },
      "to": "region=region-1,app=test-app",
    },
  ],
  "models": {
    "app=test-app": {
      "className": "App",
      "model": {
        "name": "test-app",
      },
    },
    "environment=qa,region=region-1,app=test-app": {
      "className": "Environment",
      "model": {
        "environmentName": "qa",
        "environmentVariables": {},
      },
    },
    "region=region-1,app=test-app": {
      "className": "Region",
      "model": {
        "regionId": "region-1",
      },
    },
  },
  "version": "v0",
}
`;

exports[`Serialization Service UT serialize() should serialize when multiple models have dependency on same model 1`] = `
{
  "dependencies": [
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "name",
          "toField": "imageId",
        },
      ],
      "from": "app=test-app",
      "relationship": {
        "onField": "name",
        "toField": "imageId",
        "type": "parent",
      },
      "to": "image=backend-runner:0.0.1,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "name",
          "toField": "serverKey",
        },
      ],
      "from": "app=test-app",
      "relationship": {
        "onField": "name",
        "toField": "serverKey",
        "type": "parent",
      },
      "to": "server=backend,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "serverKey",
          "toField": "name",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "serverKey",
          "toField": "name",
        },
      ],
      "from": "server=backend,app=test-app",
      "relationship": {
        "onField": "serverKey",
        "toField": "name",
        "type": "child",
      },
      "to": "app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "serverKey",
          "toField": "deploymentTag",
        },
      ],
      "from": "server=backend,app=test-app",
      "relationship": {
        "onField": "serverKey",
        "toField": "deploymentTag",
        "type": "parent",
      },
      "to": "deployment=0.0.1,server=backend,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "serverKey",
          "toField": "deploymentTag",
        },
      ],
      "from": "server=backend,app=test-app",
      "relationship": {
        "onField": "serverKey",
        "toField": "deploymentTag",
        "type": "parent",
      },
      "to": "deployment=0.0.2,server=backend,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "deploymentTag",
          "toField": "imageId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "deploymentTag",
          "toField": "imageId",
        },
      ],
      "from": "deployment=0.0.2,server=backend,app=test-app",
      "relationship": undefined,
      "to": "image=backend-runner:0.0.1,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "deploymentTag",
          "toField": "serverKey",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "deploymentTag",
          "toField": "serverKey",
        },
      ],
      "from": "deployment=0.0.2,server=backend,app=test-app",
      "relationship": {
        "onField": "deploymentTag",
        "toField": "serverKey",
        "type": "child",
      },
      "to": "server=backend,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "imageId",
          "toField": "deploymentTag",
        },
      ],
      "from": "image=backend-runner:0.0.1,app=test-app",
      "relationship": undefined,
      "to": "deployment=0.0.1,server=backend,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "imageId",
          "toField": "deploymentTag",
        },
      ],
      "from": "image=backend-runner:0.0.1,app=test-app",
      "relationship": undefined,
      "to": "deployment=0.0.2,server=backend,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "imageId",
          "toField": "name",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "imageId",
          "toField": "name",
        },
      ],
      "from": "image=backend-runner:0.0.1,app=test-app",
      "relationship": {
        "onField": "imageId",
        "toField": "name",
        "type": "child",
      },
      "to": "app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "deploymentTag",
          "toField": "imageId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "deploymentTag",
          "toField": "imageId",
        },
      ],
      "from": "deployment=0.0.1,server=backend,app=test-app",
      "relationship": undefined,
      "to": "image=backend-runner:0.0.1,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "deploymentTag",
          "toField": "serverKey",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "deploymentTag",
          "toField": "serverKey",
        },
      ],
      "from": "deployment=0.0.1,server=backend,app=test-app",
      "relationship": {
        "onField": "deploymentTag",
        "toField": "serverKey",
        "type": "child",
      },
      "to": "server=backend,app=test-app",
    },
  ],
  "models": {
    "app=test-app": {
      "className": "App",
      "model": {
        "name": "test-app",
      },
    },
    "deployment=0.0.1,server=backend,app=test-app": {
      "className": "Deployment",
      "model": {
        "deploymentTag": "0.0.1",
        "image": {
          "context": "image=backend-runner:0.0.1,app=test-app",
        },
      },
    },
    "deployment=0.0.2,server=backend,app=test-app": {
      "className": "Deployment",
      "model": {
        "deploymentTag": "0.0.2",
        "image": {
          "context": "image=backend-runner:0.0.1,app=test-app",
        },
      },
    },
    "image=backend-runner:0.0.1,app=test-app": {
      "className": "Image",
      "model": {
        "dockerOptions": {
          "buildArgs": {},
          "dockerFilePath": "/Dockerfile",
          "quiet": undefined,
        },
        "imageId": "0.0.1",
        "imageName": "backend-runner",
        "imageTag": "0.0.1",
      },
    },
    "server=backend,app=test-app": {
      "className": "Server",
      "model": {
        "serverKey": "backend",
      },
    },
  },
  "version": "v0",
}
`;
